### PR TITLE
Adds metrics model

### PR DESCRIPTION
**Pull request by suryabulusu**
- Frontend metrics are received in backend and sent to mongoDB
- Added to sessions collection document

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -73,6 +73,44 @@ class Event(BaseModel):
...
+    wronglyAnswered: int
+    partiallyAnswered: int
+    attemptRate: float
+    accuracyRate: float
+
+
+class QuestionSetMetric(BaseModel):
+    name: str
+    marks_scored: int
+    max_questions_allowed_to_attempt: int
```
Comment (deepansh96) [inline app/models.py:92]: Why is this needed from frontend?

File: app/models.py [hunk 2]
```diff
@@ -73,6 +73,44 @@ class Event(BaseModel):
...
+    num_skipped: int
+    num_correct: int
+    num_wrong: int
+    num_partially_correct: int
+    attempt_rate: float
+    accuracy_rate: float
+
+
+class SessionMetrics(BaseModel):
+    total_sets: int
```
Comment (deepansh96) [inline app/models.py:103]: what is sets here? question sets? what does this metric represent about a session?

File: app/models.py [hunk 3]
```diff
@@ -423,6 +461,7 @@ class Session(BaseModel):
     created_at: datetime = Field(default_factory=datetime.utcnow)
     events: List[Event] = []
     has_quiz_ended: bool = False
+    metrics: Optional[SessionMetrics] = None
```
Comment (deepansh96) [inline app/models.py:464]: why is this needed? when session starts there will be no metrics to report right?

File: app/routers/sessions.py [hunk 4]
```diff
@@ -117,6 +170,7 @@ async def create_session(session: Session):
         current_session["events"] = last_session.get("events", [])
         current_session["time_remaining"] = last_session.get("time_remaining", None)
         current_session["has_quiz_ended"] = last_session.get("has_quiz_ended", False)
+        current_session["metrics"] = last_session.get("metrics", None)
```
Comment (deepansh96) [inline app/routers/sessions.py:173]: so we're carrying over the metrics across sessions?

File: app/routers/sessions.py [hunk 5]
```diff
@@ -274,10 +330,17 @@ async def update_session(session_id: str, session_updates: UpdateSession):
 
     # update the document in the sessions collection
     if new_event == EventType.end_quiz:
+        session_metrics = reformat_metrics(jsonable_encoder(session_updates)["metrics"])
```
Comment (deepansh96) [inline app/routers/sessions.py:333]: one thing I didn't understand is why are we doing this reformatting / extra processing. Isn't everything already calculated on the frontend and should be just sent to the DB through the backend?
Also worried about these extra processing because if its unnecessary computation then it will put load on the system when its under scale. Either if its lambda or ec2 server.

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: app/models.py (hunk 1 above)
Comment (suryabulusu) [reply app/models.py:92]: i will remove it

File: app/models.py (hunk 2 above)
Comment (suryabulusu) [reply app/models.py:103]: no need to remove it

File: app/routers/sessions.py (hunk 5 above)
Comment (suryabulusu) [reply app/routers/sessions.py]: removed reformatting. shifted to frontend

**Review by deepansh96 — APPROVED** [review]
